pkg/apis/metal3/v1alpha1/baremetalhost_types.go

```diff
@@ -853,6 +857,50 @@ func (host *BareMetalHost) GetImageChecksum() (string, string, bool) {
 	}
 }
 
+// ValidateForAdmission returns an error if the host fails uniqueness validation,
+// and cannot be added to the cluster.  This is ordinarily run by an
+// admission webhook.
+func (host *BareMetalHost) ValidateForAdmission(c client.Client) error {
+	ctx := context.Background()
+	hostList := BareMetalHostList{}
+	opts := &client.ListOptions{
+		Namespace: host.Namespace,
+	}
+
+	err := c.List(ctx, &hostList, opts)
```

Comment:
Can you use field selectors to do a query for existing BMH with the same MAC or BMC address instead of listing all the hosts (of which there could be thousands) e.g. https://godoc.org/sigs.k8s.io/controller-runtime/pkg/client#ListOptions

Not sure if the API tools work that way to allow us to do such a query (or if it's even better)

Reply:
> Can you use field selectors to do a query for existing BMH with the same MAC or BMC address instead of listing all the hosts (of which there could be thousands)?

No. The field selector only operates on a small number of fields, think `metadata.name`.

There seem to be many fields we need to check, and doing a query for each one seems slow. This way we grab all of the resources at once, and can quickly run through them to validate them.

Reply:
Is the client used the same type of caching client that a controller uses?